proxy/src/timeout.rs

```diff
@@ -0,0 +1,179 @@
+    type Future = TimeoutFuture<S::Future>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.inner.poll_ready().map_err(Self::Error::from)
+    }
+
+    fn call(&mut self, req: Self::Request) -> Self::Future {
+        let duration = self.duration;
+        // TODO: should this panic or wrap the error?
+        let timeout = ReactorTimeout::new(duration, &self.handle)
+            .expect("failed to create timeout!");
+        let inner = self.inner.call(req);
+        TimeoutFuture {
+            inner,
+            duration,
+            timeout,
+        }
+    }
+}
+
+
+impl<C> Connect for Timeout<C>
+where
+    C: Connect,
+    // C::Error: Error,
+{
+    type Connected = C::Connected;
+    type Error = TimeoutError<C::Error>;
+    type Future = TimeoutFuture<C::Future>;
+
+    fn connect(&self) -> Self::Future {
+        let duration = self.duration;
+        // TODO: should this panic or wrap the error?
+        let timeout = ReactorTimeout::new(duration, &self.handle)
+            .expect("failed to create timeout!");
+        let inner = self.inner.connect();
+        TimeoutFuture {
+            inner,
+            duration,
+            timeout,
+        }
+    }
+}
+
+
+//===== impl TimeoutError =====
+
+impl<E> From<E> for TimeoutError<E> {
+    #[inline] fn from(error: E) -> Self {
+        TimeoutError::Error(error)
+    }
+}
+
+impl<E> fmt::Display for TimeoutError<E>
+where
+    E: fmt::Display
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            TimeoutError::Timeout(ref duration) => 
+                // TODO: format the duration nicer.
+                write!(f, "operation timed out after {:?}", duration),
+            TimeoutError::Error(ref err) =>
+                write!(f, "inner operation failed: {}", err),
+        }
+    }
+}
+
+impl<E> Error for TimeoutError<E> 
+where 
+    E: Error
+{ 
+    fn cause(&self) -> Option<&Error> {
+        match *self {
+            TimeoutError::Error(ref err) => Some(err),
+            _ => None,
+        }
+    }
+
+    fn description(&self) -> &str {
+        match *self {
+            TimeoutError::Timeout(_) => "operation timed out",
+            TimeoutError::Error(ref err) => err.description(),
+        }
+    }
+}
+
+//===== impl TimeoutFuture =====
+
+impl<F> Future for TimeoutFuture<F>
+where
+    F: Future,
+    // F::Error: Error,
+{
+    type Item = F::Item;
+    type Error = TimeoutError<F::Error>;
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        if let Async::Ready(item) = self.inner.poll().map_err(TimeoutError::from)? {
+            Ok(Async::Ready(item))
+        } else if let Async::Ready(_) = self.timeout.poll().expect("timer failed") {
+            Err(TimeoutError::Timeout(self.duration))
+        } else {
+            Ok(Async::NotReady)
+        } 
+    }
+}
+
+// We have to provide a custom implementation of Debug, because
+// tokio_core::reactor::Timeout is not Debug.
```

Comment:
We really need to stop doing this. What's stopping the new new version of `tokio_core` that includes this fix?